sky/backends/cloud_vm_ray_backend.py

```diff
@@ -2257,58 +2289,73 @@ def update_cluster_ips(
             exceptions.FetchIPError: if we failed to get the IPs. e.reason is
                 HEAD or WORKER.
         """
-
-        def is_provided_ips_valid(ips: Optional[List[Optional[str]]]) -> bool:
-            return (ips is not None and
-                    len(ips) == self.num_ips_per_node * self.launched_nodes and
-                    all(ip is not None for ip in ips))
-
-        use_internal_ips = self._use_internal_ips()
-
-        # cluster_feasible_ips is the list of IPs of the nodes in the cluster
-        # which can be used to connect to the cluster. It is a list of external
-        # IPs if the cluster is assigned public IPs, otherwise it is a list of
-        # internal IPs.
-        cluster_feasible_ips: List[str]
-        if is_provided_ips_valid(external_ips):
-            logger.debug(f'Using provided external IPs: {external_ips}')
-            cluster_feasible_ips = typing.cast(List[str], external_ips)
+        self.cached_cluster_info = cluster_info
+        if self.cached_cluster_info is None:
+            self._update_cluster_info()
+        if self.cached_cluster_info is not None:
+            use_internal_ips = self._use_internal_ips()
+            cluster_feasible_ips = self.cached_cluster_info.get_feasible_ips(
+                use_internal_ips)
+            cluster_internal_ips = self.cached_cluster_info.get_feasible_ips(
+                force_internal_ips=True)
```

Comment:
nit: maybe return here and dedent all the following. I'm happy w/the current impl as well, but it will help to review the git diff ;)

Reply:
This cannot be returned and dedented, as we have a shared code block that need to be executed after either if or else block finished.